pop-api/examples/fungibles/lib.rs

```diff
@@ -1,108 +1,261 @@
 mod fungibles {
 	use super::*;
 
 	#[ink(storage)]
-	#[derive(Default)]
-	pub struct Fungibles;
+	pub struct Fungibles {
+		id: TokenId,
+	}
 
 	impl Fungibles {
+		/// Instantiate the contract and wrap around an existing token.
+		///
+		/// # Parameters
+		/// * - `token` - The token.
 		#[ink(constructor, payable)]
-		pub fn new() -> Self {
-			Default::default()
+		pub fn existing(id: TokenId) -> Result<Self, PSP22Error> {
+			// Make sure token exists.
+			if !api::token_exists(id).unwrap_or_default() {
+				return Err(PSP22Error::Custom(String::from("Unknown")));
+			}
+			let contract = Self { id };
+			Ok(contract)
 		}
 
-		#[ink(message)]
-		pub fn total_supply(&self, token: TokenId) -> Result<Balance> {
-			api::total_supply(token)
+		/// Instantiate the contract and create a new token. The token identifier will be stored
+		/// in contract's storage.
+		///
+		/// # Parameters
+		/// * - `id` - The identifier of the token.
+		/// * - `admin` - The account that will administer the token.
+		/// * - `min_balance` - The minimum balance required for accounts holding this token.
+		#[ink(constructor, payable)]
+		pub fn new(id: TokenId, min_balance: Balance) -> Result<Self, PSP22Error> {
+			let contract = Self { id };
+			let contract_id = contract.env().account_id();
+			api::create(id, contract_id, min_balance).map_err(PSP22Error::from)?;
+			contract
+				.env()
+				.emit_event(Created { id, creator: contract_id, admin: contract_id });
+			Ok(contract)
 		}
+	}
 
+	impl Psp22 for Fungibles {
+		/// Returns the total token supply.
 		#[ink(message)]
-		pub fn balance_of(&self, token: TokenId, owner: AccountId) -> Result<Balance> {
-			api::balance_of(token, owner)
+		fn total_supply(&self) -> Balance {
+			api::total_supply(self.id).unwrap_or_default()
 		}
 
+		/// Returns the account balance for the specified `owner`
 		#[ink(message)]
-		pub fn allowance(
-			&self,
-			token: TokenId,
-			owner: AccountId,
-			spender: AccountId,
-		) -> Result<Balance> {
-			api::allowance(token, owner, spender)
+		fn balance_of(&self, owner: AccountId) -> Balance {
+			api::balance_of(self.id, owner).unwrap_or_default()
 		}
 
+		/// Returns the amount which `spender` is still allowed to withdraw from `owner`
 		#[ink(message)]
-		pub fn transfer(&mut self, token: TokenId, to: AccountId, value: Balance) -> Result<()> {
-			api::transfer(token, to, value)
+		fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+			api::allowance(self.id, owner, spender).unwrap_or_default()
 		}
 
+		/// Transfers `value` amount of tokens from the caller's account to account `to`
+		/// with additional `data` in unspecified format.
 		#[ink(message)]
-		pub fn transfer_from(
+		fn transfer(
 			&mut self,
-			token: TokenId,
-			from: AccountId,
 			to: AccountId,
 			value: Balance,
 			_data: Vec<u8>,
-		) -> Result<()> {
-			api::transfer_from(token, from, to, value)
+		) -> Result<(), PSP22Error> {
+			let caller = self.env().caller();
```

Comment:
Should this be `self.env().account_id()`? Because we never accept an actual caller to dispatch, the caller must be set to a contract address. However, if we do that, I don't find any use case for this PSP22 contract to be used in production.

Reply:
Tackled in https://github.com/r0gue-io/pop-node/pull/301